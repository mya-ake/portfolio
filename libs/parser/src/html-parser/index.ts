export * from './html-parser';
export * from './utils';
