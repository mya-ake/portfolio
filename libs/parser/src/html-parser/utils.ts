import type { Node, TagNode, TextNode } from './html-parser';

export const isTagNode = (node: Node): node is TagNode => {
  return node.nodeType === 'tag';
};

export const isTextNode = (node: Node): node is TextNode => {
  return node.nodeType === 'text';
};

export const findTargetTagNode = (
  targetTagName: string,
  nodes: Node[],
): TagNode | undefined => {
  return nodes.filter(isTagNode).find((node) => {
    return node.tagName === targetTagName;
  });
};

export const hasTargetTagNode = (
  targetTagName: string,
  nodes: Node[],
): boolean => {
  return !!findTargetTagNode(targetTagName, nodes);
};
