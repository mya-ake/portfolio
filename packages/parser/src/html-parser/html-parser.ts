/* eslint-disable @typescript-eslint/no-explicit-any */
import parse5 from 'parse5';

type NodeType = 'tag' | 'text';
type Attributes = Record<string, string>;

export type TagNode = {
  nodeType: 'tag';
  tagName: string;
  attrs: Attributes;
  childNodes: Node[];
};

export type TextNode = {
  nodeType: 'text';
  content: string;
};

export type Node = TagNode | TextNode;

const getNodeType = (nodeName: string): NodeType => {
  switch (nodeName) {
    case '#text':
      return 'text';
    default:
      return 'tag';
  }
};

const createTagNode = (childNode: parse5.Element): TagNode => ({
  nodeType: 'tag',
  tagName: childNode.tagName,
  attrs: childNode.attrs.reduce<Attributes>((attrs, attrItem) => {
    attrs[attrItem.name] = attrItem.value;
    return attrs;
  }, {}),
  childNodes: convertNodes((childNode as any).childNodes ?? []),
});

const createTextNode = (childNode: parse5.TextNode): TextNode => ({
  nodeType: 'text',
  content: childNode.value,
});

const convertNode = (childNode: parse5.ChildNode): Node => {
  const nodeType = getNodeType(childNode.nodeName);
  switch (nodeType) {
    case 'tag':
      return createTagNode(childNode as parse5.Element);
    case 'text':
      return createTextNode(childNode as parse5.TextNode);
  }
};

const convertNodes = (childNodes: parse5.ChildNode[]): Node[] =>
  childNodes.map(convertNode);

export const parseHtml = (html: string): Node[] => {
  const documentFragment = parse5.parseFragment(html);
  return convertNodes(documentFragment.childNodes);
};
