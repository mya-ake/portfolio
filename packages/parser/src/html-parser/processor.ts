export const removeBackslashOfHTMLAttribute = (html: string): string =>
  html.replace(/\\"/g, '"');
