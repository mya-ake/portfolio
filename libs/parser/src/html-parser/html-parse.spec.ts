import { describe, it, expect } from 'vitest';
import { parseHtml } from './html-parser';

describe('parseHtml', () => {
  describe('single root', () => {
    it('no attrs', () => {
      const result = parseHtml(`<p>text</p>`);
      expect(result).toEqual([
        {
          nodeType: 'tag',
          tagName: 'p',
          attrs: {},
          childNodes: [
            {
              nodeType: 'text',
              content: 'text',
            },
          ],
        },
      ]);
    });

    it('has attrs', () => {
      const result = parseHtml(
        `<button type="button" disabled aria-pressed="false">text</button>`,
      );
      expect(result).toEqual([
        {
          nodeType: 'tag',
          tagName: 'button',
          attrs: {
            type: 'button',
            disabled: '',
            'aria-pressed': 'false',
          },
          childNodes: [
            {
              nodeType: 'text',
              content: 'text',
            },
          ],
        },
      ]);
    });
  });

  describe('multi root', () => {
    it('', () => {
      const result = parseHtml(
        `<p class="p p1">text 1</p><p class="p">text 2</p>`,
      );
      expect(result).toEqual([
        {
          nodeType: 'tag',
          tagName: 'p',
          attrs: {
            class: 'p p1',
          },
          childNodes: [
            {
              nodeType: 'text',
              content: 'text 1',
            },
          ],
        },
        {
          nodeType: 'tag',
          tagName: 'p',
          attrs: {
            class: 'p',
          },
          childNodes: [
            {
              nodeType: 'text',
              content: 'text 2',
            },
          ],
        },
      ]);
    });
  });

  describe('deep children', () => {
    it('list', () => {
      const result = parseHtml(`<ul><li>item 1</li><li>item 1</li></ul>`);
      expect(result).toMatchSnapshot();
    });

    it('some content', () => {
      const result = parseHtml(
        `<div>no wrapped<p>item 1</p><img src="/static/sample.png" alt=""/></div>`,
      );
      expect(result).toMatchSnapshot();
    });
  });

  describe('web components', () => {
    it('list', () => {
      const result = parseHtml(`<custom-p>text</custom-p>`);
      expect(result).toMatchSnapshot();
    });
  });

  describe(`include '\\'`, () => {
    it(`with '"'`, () => {
      const result = parseHtml(
        `<a href=\\"https://example.com\\" rel=\\"noopener noreferrer\\">https://example.com</a>`,
      );
      expect(result).toEqual([
        {
          nodeType: 'tag',
          tagName: 'a',
          attrs: {
            href: 'https://example.com',
            rel: 'noopener noreferrer',
          },
          childNodes: [
            {
              nodeType: 'text',
              content: 'https://example.com',
            },
          ],
        },
      ]);
    });

    it(`with '"' and has '\\' text`, () => {
      const result = parseHtml(
        `<a href=\\"https://example.com\\" rel=\\"noopener noreferrer\\">\\</a>`,
      );
      expect(result).toEqual([
        {
          nodeType: 'tag',
          tagName: 'a',
          attrs: {
            href: 'https://example.com',
            rel: 'noopener noreferrer',
          },
          childNodes: [
            {
              nodeType: 'text',
              content: '\\',
            },
          ],
        },
      ]);
    });
  });
});
