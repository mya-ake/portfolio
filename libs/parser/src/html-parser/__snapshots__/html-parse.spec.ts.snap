// Vitest Snapshot v1

exports[`parseHtml > deep children > list 1`] = `
[
  {
    "attrs": {},
    "childNodes": [
      {
        "attrs": {},
        "childNodes": [
          {
            "content": "item 1",
            "nodeType": "text",
          },
        ],
        "nodeType": "tag",
        "tagName": "li",
      },
      {
        "attrs": {},
        "childNodes": [
          {
            "content": "item 1",
            "nodeType": "text",
          },
        ],
        "nodeType": "tag",
        "tagName": "li",
      },
    ],
    "nodeType": "tag",
    "tagName": "ul",
  },
]
`;

exports[`parseHtml > deep children > some content 1`] = `
[
  {
    "attrs": {},
    "childNodes": [
      {
        "content": "no wrapped",
        "nodeType": "text",
      },
      {
        "attrs": {},
        "childNodes": [
          {
            "content": "item 1",
            "nodeType": "text",
          },
        ],
        "nodeType": "tag",
        "tagName": "p",
      },
      {
        "attrs": {
          "alt": "",
          "src": "/static/sample.png",
        },
        "childNodes": [],
        "nodeType": "tag",
        "tagName": "img",
      },
    ],
    "nodeType": "tag",
    "tagName": "div",
  },
]
`;

exports[`parseHtml > web components > list 1`] = `
[
  {
    "attrs": {},
    "childNodes": [
      {
        "content": "text",
        "nodeType": "text",
      },
    ],
    "nodeType": "tag",
    "tagName": "custom-p",
  },
]
`;
